keps/draft-20190516-KUDO-package-manager.md

```diff
@@ -0,0 +1,177 @@
+
+### Implementation Details/Notes/Constraints
+
+We need to make sure that Packages can be easily accessed even if the underlying Storage Backend changed. Therefor, adding a middle layer like a registry server that has multiple Storage Backends as a swappable engine underneath seems like a good pattern. This also would enable users to keep track or restrict access to specific Frameworks.
+
+The idea would be a HTTP Server that can be easily accessed, even in air-gapped environments, depending on the user needs. As Storage Backends there could be a multitude such as `Local`, `Google`, `S3`, `Github`, `Minio`, `Docker` and so on.
+Having as an API a simple HTTP Server that just serves essentially a single yaml file that holds all the information also would enable us as a first iteration to use any HTTP server. This should work as it already with Google Cloud Storage or S3, all it requires is having access to the yaml file and something like the KUDO registry that serves like a middle layer inbetween could be easily added later on.
+
+Some caveats to this could for instance be defining a clean interface that won't break things when a user attempts to access Frameworks and hosted repos. Another caveat is deciding on the right proper structure in how a Package and its higher repo structure will look like. Design decisions here could potentialy impact future implementations of Storage Backends we haven't thought of yet. Another caveat would be identifying the right approach in versioning as this dictates a lot also how the structure will look like. Versioning for the `.tgz` files and Frameworks should be in SemVer.
+
+The Repo structure on your local laptop could look in the short term as follow:
+
+```bash
+/index.yaml
+/kafka
+/kafka/2.2.0
+/kafka/2.2.0/kafka-framework.yaml
+/kafka/2.2.0/kafka-frameworkversion.yaml
+/kafka/2.2.0/kafka-instance.yaml
+/kafka/2.2.0/metadata.yaml
+/zookeeper
+...
+```
+
+* This structure serves as the local repo, cache and source of truth
+* This also would be the on Github hosted official repo
+
+In the long term it would look more like KEP-0009 conform e.g.:
+
+```bash
+/index.yaml
+/kafka
+/kafka/2.2.0
+/kafka/2.2.0/framework.yaml
+/kafka/2.2.0/params.yaml
+/kafka/2.2.0/common
+/kafka/2.2.0/common/common.yaml
+/kafka/2.2.0/templates
+/kafka/2.2.0/templates/deployment.yaml
+/kafka/2.2.0/templates/...
+/zookeeper
+...
+```
+
+That would have the advantage that for distribution the opinionated structure within the `.tgz` file is not so much of importance and can be subject to change without breaking other assumptions.
+
+The `/kafka/2.2.0` folder (with whatever underlying structure) would be zipped to `kafka-2.2.0.tgz`, where `2.2.0` is the current SemVer version of the package. This is then made available through any HTTP Server (or later KUDO Registry with any Storage Backend), e.g. like on Google Cloud Storage:
+
+```bash
+/index.yaml
+/kafka-0.1.0.tgz
```

Comment:
so the storage will be flat? No `/kafka/kafka-2.2.0.tgz`? I would naturally nest frameworks in folders but I am not sure if one of the solutions has some bigger performance implications... 🤔

Reply:
I guess the flat structure is harder to mess up and easier to set up. I am fine with that 👍

Reply:
Yes the flat structure has multiple benefits and the strongest for me having it probably very compatible with many storage backends.

Reply:
:+1: